examples/factory/tests/tests.rs

```diff
@@ -0,0 +1,231 @@
+        })
+    );
+
+    // Create and produce a second product (from a different account), and check the
+    // result.
+    let product2 = create_product(&mut chain, module_ref, BOB);
+    call_produce(&mut chain, factory, product2, BOB).expect("Produce successful");
+    let prod_state = view_product(&chain, product2).expect("Get view of initialized product");
+    assert_eq!(
+        prod_state,
+        ProductState::Initialized(Product {
+            factory,
+            index: 1,
+        })
+    );
+
+    // Check that the factory records having 2 products, and that they match what we
+    // expect.
+    assert_eq!(view_product_count(&chain, factory).expect("View product count successful"), 2);
+    assert_eq!(view_product_at(&chain, factory, 0).expect("View of product at index"), product);
+    assert_eq!(view_product_at(&chain, factory, 1).expect("View of product at index"), product2);
+}
+
+/// Test that a product instance cannot be produced using an account that did
+/// not create the product instance.
+#[test]
+fn test_not_authorized() {
+    let (mut chain, module_ref) = initialize_chain();
+    let factory = create_factory(&mut chain, module_ref);
+    // ALICE creates the product instance, but BOB invokes produce on the factory.
+    // This should fail.
+    let product = create_product(&mut chain, module_ref, ALICE);
+    call_produce(&mut chain, factory, product, BOB).expect_err(
+        "Should fail when product is called from a different account than the creator of the \
+         product.",
+    );
+}
+
+/// Test that a product cannot be produced twice (by the same factory).
+#[test]
+fn test_double_produce() {
+    let (mut chain, module_ref) = initialize_chain();
+    let factory = create_factory(&mut chain, module_ref);
+    let product = create_product(&mut chain, module_ref, ALICE);
+    call_produce(&mut chain, factory, product, ALICE).expect("Produce successful");
+    call_produce(&mut chain, factory, product, ALICE)
+        .expect_err("Should fail when product has already been initialized.");
+}
+
+/// Test that a product cannot be produced twice (by different factories).
+#[test]
+fn test_double_produce_second_factory() {
+    let (mut chain, module_ref) = initialize_chain();
+    let factory = create_factory(&mut chain, module_ref);
+    let product = create_product(&mut chain, module_ref, ALICE);
+    call_produce(&mut chain, factory, product, ALICE).expect("Produce successful");
+    let factory2 = create_factory(&mut chain, module_ref);
+    call_produce(&mut chain, factory2, product, ALICE)
+        .expect_err("Should fail when product has already been initialized.");
+}
+
+/// Test that `produce` fails when called with an instance of a contract other
+/// than the product.
+#[test]
+fn test_wrong_product() {
+    let (mut chain, module_ref) = initialize_chain();
+    let factory = create_factory(&mut chain, module_ref);
+    let factory2 = create_factory(&mut chain, module_ref);
```

Comment:
[nitpick] The test specifically tests this condition `product_name != PRODUCT_INIT_NAME`, which is a special case of "a contract other than the product." Since this is something relevant to the contract logic, I think it makes sense to reflect it in the documentation. It would be nice to have a test with another module reference, but it's a bit annoying, because we would need another contract for that.

Reply:
I agree. Updated the doc, but I'm not going to the effort of testing it with a whole different module.